core/iwasm/common/wasm_runtime_common.c

```diff
@@ -2342,8 +2342,8 @@ wasm_runtime_get_exec_env_singleton(WASMModuleInstanceCommon *module_inst_comm)
     return module_inst->exec_env_singleton;
 }
 
-void
-wasm_set_exception(WASMModuleInstance *module_inst, const char *exception)
+static void
+wasm_set_exception_local(WASMModuleInstance *module_inst, const char *exception)
```

Comment:
Can we remove the lock in this function?

Reply:
exception_lock() is already no-op for builds w/o threading support.

Reply:
When thread manager is enabled, wasm_set_exception_local is called only when exec_env isn't in the cluster, seems no lock is needed. But exception_lock() uses lock when thread manager is enabled.

Reply:
well, can cluster be NULL here?
even if it's possible, it's exceptional, isn't it?

Reply:
The cluster may be NULL if `wasm_exec_env_create` hasn't been called with the module_inst, or the exec_env was destroyed by `wasm_exec_env_destroy` and the cluster was destroyed.

Yes, it is exceptional, I just think the `_local` suffix means we can only care about the current module_inst, no need to care about multi-threading. It is also OK if you prefer to keep the exception_lock/unlock.

Reply:
> The cluster may be NULL if `wasm_exec_env_create` hasn't been called with the module_inst, or the exec_env was destroyed by `wasm_exec_env_destroy` and the cluster was destroyed.

ok.
i feel it's awkward to use a trap w/o exec env though. cf. https://github.com/bytecodealliance/wasm-micro-runtime/issues/2512

> Yes, it is exceptional, I just think the `_local` suffix means we can only care about the current module_inst, no need to care about multi-threading. It is also OK if you prefer to keep the exception_lock/unlock.

i prefer to keep exception_lock here.
as this is not performance critical at all, consistency outweighs IMO.